Solve Day 6, Part 1

byronic opened the pull request:
Solution for day 6, part 1; I heart 2D coordinate stuff so this was a lot of fun; remembering to branch this time so I can solicit reviews

Commit:
Solve Day 6, Part 1

Solution for day 6, part 1; I heart 2D coordinate stuff so this was a lot of fun; remembering to branch this time so I can solicit reviews

Primetalk commented on 06/Coordinatron.scala:
```diff
@@ -0,0 +1,109 @@
+
+  // finds the bounds as follows: x min, x max, y min, y max
+  def bounds(l: List[(Int, Int)]): (Int, Int, Int, Int) = {
+    @annotation.tailrec
```
👍

Primetalk commented on 06/Coordinatron.scala:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  // returns the list's closest point (or one of them), and true if equidistant to multiple points or false otherwise
+  def findClosestPoint(p: (Int, Int), l: List[(Int, Int)]): ((Int, Int), Boolean) = {
```
This could be `Option[(Int, Int)]` as well.

Primetalk commented on 06/Coordinatron.scala:
```diff
@@ -0,0 +1,109 @@
+  // returns the list's closest point (or one of them), and true if equidistant to multiple points or false otherwise
+  def findClosestPoint(p: (Int, Int), l: List[(Int, Int)]): ((Int, Int), Boolean) = {
+    def go(p: (Int, Int), l: List[(Int, Int)], s: Int, c: (Int, Int), e: Boolean): ((Int, Int), Boolean) = {
+      if (l.isEmpty) (c, e)
```
You might find pattern matching more convenient for recursive functions:
```scala
l match {
 case Nil =>
 case h :: t => 
}
```

coacoas commented on 06/Coordinatron.scala:
```diff
@@ -0,0 +1,109 @@
+object Coordinatron {
+  val bufferedSource = Source.fromFile("dataset")
+  val lines = bufferedSource.getLines.toList
+  bufferedSource.close
```
I would put this inside `main` so that it's only run at program startup. Having it here puts it in the constructor for the object, so it will run even during tests.

coacoas commented on 06/Coordinatron.scala:
```diff
@@ -0,0 +1,109 @@
+
+  def absoluteValue(n: Int): Int = {
+    if (n < 0) n * -1 else n
+  }
```
You can use `Int::abs` for this: 
```
scala> -4.abs
res4: Int = 4
```

coacoas commented on 06/Coordinatron.scala:
```diff
@@ -0,0 +1,109 @@
+
+  def max(a: Int, b: Int): Int = {
+    if (a > b) a else b
+  }
```
`Int` also has `min` and `max` methods: 
```
scala> 1.min(2)
res5: Int = 1

scala> 1.max(2)
res6: Int = 2
```

coacoas commented on 06/Coordinatron.scala:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  // finds the bounds as follows: x min, x max, y min, y max
+  def bounds(l: List[(Int, Int)]): (Int, Int, Int, Int) = {
```
Tuples are generally best used only for _very_ short lived and small constructs, especially if the values have different types. The issue here is that it's hard to remember which `Int` means which. Luckily, you can create a useful class with one line: 
```scala
case class Bounds(minX: Int, maxX: Int, minY: Int, maxY: Int)
```
And now you have names to indicate which `Int` is which, rather than `._1`, `._2`, etc.

coacoas commented on 06/Coordinatron.scala:
```diff
@@ -0,0 +1,109 @@
+
+  // returns the list's closest point (or one of them), and true if equidistant to multiple points or false otherwise
+  def findClosestPoint(p: (Int, Int), l: List[(Int, Int)]): ((Int, Int), Boolean) = {
+    def go(p: (Int, Int), l: List[(Int, Int)], s: Int, c: (Int, Int), e: Boolean): ((Int, Int), Boolean) = {
```
I know we use single-letter variable names a lot in FP, but those are for when the value is abstract/generic. In this case, I think this could benefit from more descriptive names (for instance, what do `s`, `e`, and `c` mean?

coacoas replied:
Also note that I think you are only ever passing `p` to `go`.  In that case, you can eliminate `p` from the parameters of `go` and it will reference the `p` in the external scope (the one passed to `findClosestPoint`.

coacoas commented on 06/Coordinatron.scala:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    def updateList(l: List[(Int, Int)], p: ((Int, Int), Boolean)): List[(Int, Int)] = {
+      if (p._2 || l.contains(p)) l else p._1 :: l
```
`l.contains(p)` will always return `false`, since `l` is a `List[(Int, Int)]` and `p` is a `((Int, Int), Boolean)`.

coacoas commented on 06/Coordinatron.scala:
```diff
@@ -0,0 +1,109 @@
+
+  // do a loop from all bounds minus one, calculate each point distance, then be done
+  // also, save a list of infinites (closest points to outside border are hitting infinite space)
+  def makeMap(l: List[(Int, Int)], bounds: (Int, Int, Int, Int)): (Map[(Int, Int), Int], List[(Int, Int)]) = {
```
I'm getting lost in some of these types and I'm not sure what is what. I definitely suggest using `case class` to clear up some things, and sometimes just some type aliases can show a little more... 
```scala
type X = Int
type Y = Int
type Area = Int
case class Point(x: X, y: Y)
case class Bounds(minX: X, maxX: X, minY: Y, maxY: Y)

def makeMap(
  points: List[Point],
  bounds: Bounds,
): (Map[Point, Area], List[Point])
```
Something like that?  Did I read it correctly?

coacoas commented on 06/Coordinatron.scala:
```diff
@@ -0,0 +1,109 @@
+  // do a loop from all bounds minus one, calculate each point distance, then be done
+  // also, save a list of infinites (closest points to outside border are hitting infinite space)
+  def makeMap(l: List[(Int, Int)], bounds: (Int, Int, Int, Int)): (Map[(Int, Int), Int], List[(Int, Int)]) = {
+    def updateMap(m: Map[(Int, Int), Int], p: ((Int, Int), Boolean)): Map[(Int, Int), Int] = {
```
I definitely suggest case classes here... I _think_ the `Boolean` indicates whether it's infinite or not?  Maybe another type `MapPoint` to be something like: 
```
case class MapPoint(p: Point, isInfinite: Boolean)
```

coacoas reviewed (commented):
Added a few comments. Primarily, I got a bit lost in the names and the use of tuples.  I highly recommend creating case classes for your data types - they're super cheap and tremendously valuable for communicating intent.

Commit:
Solve Day 6, Part 2

Add solution for part two of day 6; throws another couple of tail recursive functions on the fire

byronic commented:
@coacoas thanks for all the feedback! I'll see about cleaning some of this up. I wasn't aware of the absolute value / max options for integers and although I thought about the Tuples being a bit difficult to read, I thought that was a better practice. This'll get a lot more readable with some polish.

Commit:
Cleanup per @coacoas Review

Move file read to main; replace custom min, max, and absolute value functions with the ones provided by Scala; add case classes for points, points with an infinite flag, and the x/y bounding points to replace the use of unnamed and hard-to-read 2- and 4-tuples; removes unneeded point-passing inside tail-recursive functions enclosed by the point already in scope; generally attempt to make this less eye-gougingly terrible to read

coacoas commented on 06/Coordinatron.scala:
```diff
@@ -0,0 +1,120 @@
+  // returns the list's closest point (or one of them), and true if equidistant to multiple points or false otherwise
+  def findClosestPoint(point: Point, l: List[Point]): MapPoint = {
+    def go(l: List[Point], shortestDistance: Int, closestPoint: Point, equidistant: Boolean): MapPoint = {
+      if (l.isEmpty) MapPoint(closestPoint, equidistant)
```
A little confusing here, and this may have been my misunderstanding from before. The field in `MapPoint` is called `isInfinite`, but you are assigning a value called `equidistant`. Does this field represent whether or not the distance is strictly equidistant, or is it more of a "this point doesn't count" notion?  If the latter, I would name the field something like `isValid`.  If the former, I'd rename the field to `isEquidistant`.

Primetalk commented on 06/Coordinatron.scala:
```diff
@@ -1,129 +1,119 @@
-  def min(a: Int, b: Int): Int = {
-    if (a < b) a else b
+  def distance(p1: Point, p2: Point): Int = {
+    (p1.x - p2.x).abs + (p1.y - p2.y).abs
```
👍

Primetalk commented on 06/Coordinatron.scala:
```diff
@@ -1,129 +1,119 @@
-    def go(l: List[(Int, Int)], xMin: Int, xMax: Int, yMin: Int, yMax: Int): (Int, Int, Int, Int) = {
-      if (l.isEmpty) (xMin, xMax, yMin, yMax)
-      else go(l.tail, min(l.head._1, xMin), max(l.head._1, xMax), min(l.head._2, yMin), max(l.head._2, yMax))
+    def go(l: List[Point], xMin: Int, xMax: Int, yMin: Int, yMax: Int): Bounds = {
```
The last 4 arguments look like a rectangle. You could use `rect: Bounds`.

Primetalk commented on 06/Coordinatron.scala:
```diff
@@ -1,129 +1,119 @@
-      if (l.isEmpty) (xMin, xMax, yMin, yMax)
-      else go(l.tail, min(l.head._1, xMin), max(l.head._1, xMax), min(l.head._2, yMin), max(l.head._2, yMax))
+    def go(l: List[Point], xMin: Int, xMax: Int, yMin: Int, yMax: Int): Bounds = {
+      if (l.isEmpty) Bounds(xMin, xMax, yMin, yMax)
```
You might find pattern matching convenient for recursive functions:
```scala
l match {
  case Nil => Bounds...
  case p::ps =>
    go(ps, ...)
}
```

Primetalk commented on 06/Coordinatron.scala:
```diff
@@ -1,129 +1,119 @@
-    def go(p: (Int, Int), l: List[(Int, Int)], s: Int, c: (Int, Int), e: Boolean): ((Int, Int), Boolean) = {
-      if (l.isEmpty) (c, e)
+  def findClosestPoint(point: Point, l: List[Point]): MapPoint = {
+    def go(l: List[Point], shortestDistance: Int, closestPoint: Point, equidistant: Boolean): MapPoint = {
```
👍 Nice names and type names.

Primetalk commented on 06/Coordinatron.scala:
```diff
@@ -1,129 +1,119 @@
+         equidistant must become false
+         if currentDistance equals shortestDistance, we have seen this distance before; equidistant must become true
+         in all other cases equidistant must remain at the value passed to this iteration of go */
+        val newEquidistant: Boolean = if (shorter != shortestDistance) false else if (currentDistance == shortestDistance) true else equidistant
```
Constructions like `if(a) true else false` Could be replaced with just `a`.
If you have additional conditions, you might use `||` or `&&` for collapsing boolean expressions.
In this case the last `if` could be rewritten according to the following rule:
```scala
if(a) true else b ===> a || b
```
then the first if could be replaced:
```scala
if(a) false else b ===> !a && b
```
So you might get something like
```scala
val newEquidistant = shorter == shortestDistance && (currentDistance == shortestDistance || equidistant)
```

Primetalk commented on 06/Coordinatron.scala:
```diff
@@ -1,129 +1,119 @@
+  def makeMap(l: List[Point], bounds: Bounds): (Map[Point, Int], List[Point]) = {
+    def updateMap(m: Map[Point, Int], p: MapPoint): Map[Point, Int] = {
+      if (p.isInfinite) m
+      else m.updated(p.point, m.get(p.point) match {
```
pattern matching on `Option` can often be collapsed to `fold` or `map` + `getOrElse`:
```scala
option.map(_ + 1).getOrElse(None)
option.fold(1)(_ + 1)
```

Primetalk commented on 06/Coordinatron.scala:
```diff
@@ -1,129 +1,119 @@
+      def go(ps: List[Point], acc: Int): Int = {
         if (ps.isEmpty || acc > 10000) acc
-        else go(p, ps.tail, acc + distance(p, ps.head))
+        else go(ps.tail, acc + distance(p, ps.head))
```
👍 `tailrec`

Primetalk commented on 06/Coordinatron.scala:
```diff
@@ -1,129 +1,119 @@
       }
-      println(s"Summing distances for ${p}")
-      go(p, ps, 0)
+      go(points, 0)
```
You can often find combinators that are easier to use than recursion: `points.map(distance(p, _)). sum`

Primetalk reviewed (approved):
LGTM
